smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/locals/ContextDecorator.java

```diff
@@ -0,0 +1,117 @@
+package io.smallrye.reactive.messaging.providers.locals;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+import io.smallrye.common.vertx.VertxContext;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.operators.MultiOperator;
+import io.smallrye.mutiny.operators.multi.MultiOperatorProcessor;
+import io.smallrye.mutiny.subscription.MultiSubscriber;
+import io.smallrye.reactive.messaging.providers.PublisherDecorator;
+import io.vertx.core.Context;
+import io.vertx.core.Vertx;
+
+@ApplicationScoped
+public class ContextDecorator implements PublisherDecorator {
+    @Override
+    public Multi<? extends Message<?>> decorate(Multi<? extends Message<?>> publisher, String channelName) {
+        return publisher
+                .plug(upstream -> new ContextMulti((Multi<Message<?>>) upstream));
+    }
+
+    static class ContextMulti extends MultiOperator<Message<?>, Message<?>> {
+
+        public ContextMulti(Multi<Message<?>> upstream) {
+            super(upstream);
+        }
+
+        @Override
+        public void subscribe(MultiSubscriber<? super Message<?>> subscriber) {
+            MultiOperatorProcessor<Message<?>, Message<?>> operator = new ContextProcessor(subscriber);
+            upstream().subscribe().withSubscriber(operator);
+        }
+
+        static class ContextProcessor extends MultiOperatorProcessor<Message<?>, Message<?>> {
+
+            private volatile Context rootContext;
+
+            private static final AtomicReferenceFieldUpdater<ContextProcessor, Context> ROOT_CONTEXT_UPDATER = AtomicReferenceFieldUpdater
+                    .newUpdater(ContextProcessor.class, Context.class, "rootContext");
+
+            public ContextProcessor(MultiSubscriber<? super Message<?>> downstream) {
+                super(downstream);
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                // Release context on terminal events
+                Context root = ROOT_CONTEXT_UPDATER.getAndSet(this, null);
+                // Do NOT check for the current context and trampoline, as we may short-cut the completion event
+                // before the processing of the items.
+                if (root == null) {
+                    super.onFailure(throwable);
+                } else {
+                    root.runOnContext(ignored -> super.onFailure(throwable));
```

Comment:
it works because the duplicated contexts should be on the same event loop. However, when using multiple partitions, it will not work.

Reply:
We are running only terminal events (onComplete, onFailure) on the captured root event loop. Otherwise we need to apply the Context decorator on each partition stream independently.